### ISSUE TITLE
Please consider implementing conditional moves and the FISTTP instruction

### ISSUE THREAD

**Issue by prino** [joncampbell123/dosbox-x#2526]
And yes, I know, they're P6 and SSE3, but the latter might not require too much effort (store FPU CW, set to truncate, call FISTP, restore FPU CW) and the encoding is pretty much bog-standard for the FPU instructions.

**Comment by Wengier** [comment]
Are you able to implement the instruction yourself? If so, please feel free to send a PR to implement it (anyone else can do it too). Thanks.

**Comment by joncampbell123** [comment]
Conditional moves appeared in the Pentium Pro, which would be perfectly fine to add to cputype=ppro_slow

http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/CPU/Pentium%20Pro/Pentium%20Pro%20Family%20Developer%27s%20Manual%20Volume%202%20%2d%20Programmer%e2%80%99s%20Reference%20Manual%20%281995%2d12%29%2epdf

**Comment by prino** [comment]
Actually, I completely missed that there is a PPro CPU type, which happily accepts CMOVcc instructions. As for the FISTTP, if I knew C(++) I'd be happy to have a go at it. I'll see if I can download the entice source and see where FISTP is coded, and how it interacts with the FPU control word.

**Comment by bitbanditorg** [comment]
Actually, it's pretty straightforward to implement the fisttp instruction. Following the implementation of fistp just a few lines of code required. I made a diff for /src/fpu/fpu.cpp and /src/fpu/fpu_instructions_x86.h files. Maybe extra IFs required to limit fisttp to PentiumPro: if (CPU_ArchitectureType==CPU_ARCHTYPE_PPROSLOW). Could you please include the changes to the main distribution? DOS 256b intros may require it. Thank you!

/src/fpu/fpu.cpp

385c385,386
<               LOG(LOG_FPU,LOG_WARN)("ESC 3 EA:Unhandled group %d subfunction %d",(int)group,(int)sub);
---
>               FPU_FSTT_I32(addr);
>               FPU_FPOP();
538c539,540
<               LOG(LOG_FPU,LOG_WARN)("ESC 5 EA:Unhandled group %d subfunction %d",(int)group,(int)sub);
---
>               FPU_FSTT_I64(addr);
>               FPU_FPOP();
660,661c662,664
<       case 0x01:
<               LOG(LOG_FPU,LOG_WARN)("ESC 7 EA:Unhandled group %d subfunction %d",(int)group,(int)sub);
---
>       case 0x01:   /* FISTTP int16_t */
>               FPU_FSTT_I16(addr);
>               FPU_FPOP();

/src/fpu/fpu_instructions_x86.h

1054a1055,1059
> static void FPU_FSTT_I16(PhysPt addr) {
>       FPUD_STORE(fisttp,WORD,s)
>       mem_writew(addr,(uint16_t)fpu.p_regs[8].m1);
> }
>
1059a1065,1069
> static void FPU_FSTT_I32(PhysPt addr) {
>       FPUD_STORE(fisttp,DWORD,l)
>       mem_writed(addr,fpu.p_regs[8].m1);
> }
>
1063a1074,1079
> }
>
> static void FPU_FSTT_I64(PhysPt addr) {
>       FPUD_STORE(fisttp,QWORD,q)
>       mem_writed(addr,fpu.p_regs[8].m1);
>       mem_writed(addr + 4, fpu.p_regs[8].m2);

**Comment by joncampbell123** [comment]
The thing is that FISTTP first appeared in the Pentium 4. Since DOSBox-X does not implement anything past Pentium Pro, it may be wise to add a CPU type that's a sort of "experimental" CPU type until someday Pentium 2, 3, and 4 emulation is added. cputype=experimental, perhaps? Experimental would include Pentium Pro and newer instructions.

**Comment by joncampbell123** [comment]
I just added cputype=experimental. All you have to do is add a case statement for FISTTP and make sure it goes to the Undefined Opcode exception case unless CPU_ARCHTYPE_EXPERIMENTAL.

cputype=experimental is also intended for any other instructions added after the Pentium Pro. CPUID will still report a Pentium Pro at this time though.

**Comment by bitbanditorg** [comment]
You're right, I've completely overlooked this. For some reason I believe FISTTP was added to Pentium Pro but indeed it was added along with SSE3. I've just pulled the changes and start fiddling with it.

Should I look into the dynamic core implementation? The /src/cpu/core_dyn_x86/dyn_fpu.h and /src/cpu/core_dynrec/dyn_fpu.h headers also mention the FISTTP and referenced the /src/fpu/fpu_instructions_x86.h so it seems easy to include FISTTP.
I haven't dig into the dyn_fpu_dh.h myself, it differs a lot.

I've noticed that the debugger isn't able to display all FPU instructions, ie. fucomi, fucomip. I will check the differences.
